docs/decisions/0011-kernel-service-registration.md

```diff
@@ -0,0 +1,180 @@
+
+```csharp
+public interface IKernelServiceProvider 
+{
+    T? GetService<T>(string? name = null);
+} 
+
+public interface IKernel 
+{
+    IKernelServiceProvider Services { get; }
+}
+```
+
+```csharp
+var kernel = Kernel.Builder
+    .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+    .WithOpenAITextEmbeddingGenerationService(modelId, apiKey)
+    .WithService<IMemoryStore, VolatileMemoryStore>(),
+    .WithService<ISemanticTextMemory, SemanticTextMemory>()
+    .Build();
+
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
+```
+
+Pros:
+
+- No dependency on specific DI container library.
+- Lightweight implementation.
+- Possibility to register only those services that can be used by Plugins (isolation from host application).
+- Possibility to register same interface multiple times by **name**.
+
+Cons:
+
+- Implementation and maintenance for custom DI container, instead of using already existing libraries.
+- To import Plugin, it still needs to be initialized manually to inject specific service.
+
+### Solution #2.2
+
+This solution is an improvement for last disadvantage of Solution #2.1 to handle case, when Plugin instance should be initialized manually. This will require to add new way how to import Plugin into Kernel - not with object **instance**, but with object **type**. In this case, Kernel will be responsible for `TextMemoryPlugin` initialization and injection of all required dependencies from custom service collection.
+
+```csharp
+// Instead of this
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
+
+// Use this
+kernel.ImportFunctions<TextMemoryPlugin>();
+```
+
+### Solution #3
+
+Instead of custom service collection and service provider in Kernel, use already existing DI library - `Microsoft.Extensions.DependencyInjection`.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+
+serviceCollection.AddTransient<IMemoryStore, VolatileMemoryStore>();
+serviceCollection.AddTransient<ITextEmbeddingGeneration>(
+    (serviceProvider) => new OpenAITextEmbeddingGeneration(modelId, apiKey));
+
+serviceCollection.AddTransient<ISemanticTextMemory, SemanticTextMemory>();
+
+var services = serviceCollection.BuildServiceProvider();
+
+var kernel = Kernel.Builder
+    .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+    .WithOpenAITextEmbeddingGenerationService(modelId, apiKey)
+    .WithServices(services) // Pass all registered services from host application to Kernel
+    .Build();
+
+// Plugin Import - option #1
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
```

Comment:
Do we have a definition somewhere for what we expect the "kernel" to actually be? Is it ending up just as a tuple of a few kinds of state that can be passed around together, e.g. a list of functions and a list of registered services?  If the kernel is moving to be this super lightweight thing that's just aggregating other resources, that to me suggests we definitely don't want it to be a full-blown DI container and instead just a wrapper for one.